File: results/2026-04-09_152435/tasks/16-environment-matrix-generator/typescript-bun-sonnet/generated-code/run-act-tests.ts
/**
 * Act test harness for the environment matrix generator.
 *
 * This script:
 *  1. Creates a temporary git repository containing all project files.
 *  2. Runs `act push --rm` against the workflow.
 *  3. Captures the full output and appends it to `act-result.txt`.
 *  4. Asserts that act exited with code 0.
 *  5. Parses the output and asserts on exact expected values from each fixture.
 *
 * Run with: bun run run-act-tests.ts
 */

import { execSync, spawnSync } from "child_process";
import { mkdtempSync, cpSync, writeFileSync, appendFileSync, existsSync, rmSync } from "fs";
import { join, resolve } from "path";
import { tmpdir } from "os";

// ── Configuration ──────────────────────────────────────────────────────────────

const PROJECT_DIR = resolve(__dirname);
const ACT_RESULT_FILE = join(PROJECT_DIR, "act-result.txt");

// Expected values per fixture that we'll assert in the act output.
// These are exact strings that must appear in the workflow stdout.
const EXPECTED_VALUES: Record<string, string[]> = {
  basic: [
    '"ubuntu-latest"',
    '"windows-latest"',
    '"18"',
    '"20"',
    '"fail-fast": true',
    "COMBINATION_COUNT: 4",
  ],
  "include-exclude": [
    '"fail-fast": false',
    '"max-parallel": 3',
    '"experimental": true',
    "COMBINATION_COUNT: 4",
  ],
  "size-exceeded": [
    "MATRIX_ERROR",
    "exceeds maximum",
  ],
};

// ── Helpers ────────────────────────────────────────────────────────────────────

function log(msg: string): void {
  console.log(msg);
}

function fail(msg: string): never {
  console.error(`FAIL: ${msg}`);
  process.exit(1);
}

/**
 * Copy all relevant project files into a destination directory so act has a
 * complete, self-contained repository to work from.
 */
function copyProjectFiles(destDir: string): void {
  const filesToCopy = [
    "matrix-generator.ts",
    "index.ts",
    "matrix-generator.test.ts",
    "workflow.test.ts",
    ".github",
    "fixtures",
    ".actrc",
  ];
  for (const file of filesToCopy) {
    const src = join(PROJECT_DIR, file);
    const dst = join(destDir, file);
    if (existsSync(src)) {
      cpSync(src, dst, { recursive: true });
    }
  }
}

/**
 * Initialise a git repo in dir, add all files, and create an initial commit.
 */
function initGitRepo(dir: string): void {
  const opts = { cwd: dir, stdio: "pipe" as const };
  execSync("git init", opts);
  execSync("git config user.email 'test@example.com'", opts);
  execSync("git config user.name 'Test'", opts);
  execSync("git add -A", opts);
  execSync("git commit -m 'test: initial commit for act run'", opts);
}

/**
 * Run `act push --rm` in dir, return { exitCode, output }.
 */
function runAct(dir: string): { exitCode: number; output: string } {
  log("  Running act push --rm ...");
  const result = spawnSync(
    "act",
    ["push", "--rm", "--pull=false"],
    {
      cwd: dir,
      encoding: "utf-8",
      timeout: 300_000, // 5 min max
    }
  );
  const output = (result.stdout ?? "") + (result.stderr ?? "");
  return { exitCode: result.status ?? 1, output };
}

// ── Main test runner ───────────────────────────────────────────────────────────

async function main(): Promise<void> {
  // Initialise (or truncate) act-result.txt
  writeFileSync(ACT_RESULT_FILE, `# act-result.txt — Environment Matrix Generator\n\n`);
  log("Starting act integration test...\n");

  // We run a single act invocation with all fixtures included.
  // The workflow processes all three fixtures in sequence, so one run is enough
  // to verify all expected values.
  const tmpDir = mkdtempSync(join(tmpdir(), "matrix-gen-act-"));
  log(`Temp directory: ${tmpDir}`);

  try {
    log("  Copying project files...");
    copyProjectFiles(tmpDir);

    log("  Initialising git repository...");
    initGitRepo(tmpDir);

    appendFileSync(
      ACT_RESULT_FILE,
      `\n${"=".repeat(60)}\n` +
      `TEST CASE: all-fixtures (single act run)\n` +
      `${"=".repeat(60)}\n`
    );

    const { exitCode, output } = runAct(tmpDir);

    appendFileSync(ACT_RESULT_FILE, output);
    appendFileSync(ACT_RESULT_FILE, `\nEXIT_CODE: ${exitCode}\n`);

    log(`  act exit code: ${exitCode}`);

    // ── Assert exit code 0 ───────────────────────────────────────────────────
    if (exitCode !== 0) {
      fail(
        `act exited with code ${exitCode}. Check act-result.txt for details.`
      );
    }

    // ── Assert "Job succeeded" appeared ──────────────────────────────────────
    if (!output.includes("Job succeeded")) {
      fail('Expected "Job succeeded" in act output but it was not found.');
    }
    log('  [PASS] "Job succeeded" found in output');

    // ── Assert fixture-specific expected values ───────────────────────────────
    for (const [fixture, expectedStrings] of Object.entries(EXPECTED_VALUES)) {
      log(`\n  Checking fixture: ${fixture}`);
      for (const expected of expectedStrings) {
        if (!output.includes(expected)) {
          fail(
            `Fixture '${fixture}': expected to find '${expected}' in act output but it was not present.\n` +
            `Check act-result.txt for the full output.`
          );
        }
        log(`    [PASS] found: ${JSON.stringify(expected)}`);
      }
    }

    log("\n✓ All assertions passed!");
    log(`act-result.txt written to: ${ACT_RESULT_FILE}`);
  } finally {
    // Clean up temp directory
    try {
      rmSync(tmpDir, { recursive: true, force: true });
    } catch {
      // Best-effort cleanup
    }
  }
}

main().catch((err) => {
  console.error("Unexpected error:", err);
  process.exit(1);
});

File: results/2026-04-17_004319/tasks/12-pr-label-assigner/default-sonnet46-1m-medium/generated-code/run_act_tests.py
#!/usr/bin/env python3
"""
Act test harness for PR Label Assigner.

For each test case:
  1. Creates a temp git repo containing all project files.
  2. Writes the test case's fixture to fixtures/test_input.json.
  3. Runs `act push --rm` and captures output.
  4. Appends output (with delimiter) to act-result.txt.
  5. Asserts exit code 0, "Job succeeded", and exact ASSIGNED_LABELS value.
"""

import json
import os
import shutil
import subprocess
import sys
import tempfile

# ---------------------------------------------------------------------------
# Test cases: each defines the fixture input and the exact expected output.
# The workflow reads fixtures/test_input.json and prints ASSIGNED_LABELS.
# ---------------------------------------------------------------------------
TEST_CASES = [
    {
        "name": "docs_files_get_documentation_label",
        "input": {
            "files": ["docs/README.md", "docs/contributing.md"],
            "rules": [
                {"pattern": "docs/**",  "label": "documentation", "priority": 1},
                {"pattern": "src/**",   "label": "backend",       "priority": 2},
            ],
        },
        # Only docs files changed; no backend files -> only documentation
        "expected": "documentation",
    },
    {
        "name": "api_files_get_api_and_backend_labels",
        "input": {
            "files": ["src/api/users.py", "src/api/auth.py"],
            "rules": [
                {"pattern": "src/api/**", "label": "api",           "priority": 1},
                {"pattern": "src/**",     "label": "backend",       "priority": 2},
                {"pattern": "docs/**",    "label": "documentation", "priority": 3},
            ],
        },
        # src/api/** matches -> api; src/** matches -> backend; docs/** no match
        "expected": "api,backend",
    },
    {
        "name": "test_file_and_docs_get_tests_and_documentation",
        "input": {
            "files": ["src/api/users.test.py", "docs/README.md"],
            "rules": [
                {"pattern": "*.test.*", "label": "tests",         "priority": 1},
                {"pattern": "docs/**",  "label": "documentation", "priority": 2},
            ],
        },
        # *.test.* matches users.test.py -> tests; docs/** matches README.md -> documentation
        "expected": "tests,documentation",
    },
]

PROJECT_DIR = os.path.dirname(os.path.abspath(__file__))
ACT_RESULT_FILE = os.path.join(PROJECT_DIR, "act-result.txt")


def setup_temp_repo(tmpdir: str, fixture_input: dict) -> None:
    """Copy project files into tmpdir and write the fixture for this test case."""
    # Copy all project files except .git
    for item in sorted(os.listdir(PROJECT_DIR)):
        if item == ".git":
            continue
        src = os.path.join(PROJECT_DIR, item)
        dst = os.path.join(tmpdir, item)
        if os.path.isdir(src):
            shutil.copytree(src, dst)
        else:
            shutil.copy2(src, dst)

    # Overwrite fixtures/test_input.json with this test case's data
    fixtures_dir = os.path.join(tmpdir, "fixtures")
    os.makedirs(fixtures_dir, exist_ok=True)
    with open(os.path.join(fixtures_dir, "test_input.json"), "w") as fh:
        json.dump(fixture_input, fh, indent=2)

    # Initialise git repo
    for cmd in [
        ["git", "init"],
        ["git", "config", "user.email", "test@example.com"],
        ["git", "config", "user.name", "Test"],
        ["git", "add", "-A"],
        ["git", "commit", "-m", "test fixture"],
    ]:
        subprocess.run(cmd, cwd=tmpdir, check=True, capture_output=True)


def run_act(tmpdir: str) -> subprocess.CompletedProcess:
    """Run `act push --rm` in tmpdir and return the result."""
    return subprocess.run(
        ["act", "push", "--rm", "--pull=false"],
        cwd=tmpdir,
        capture_output=True,
        text=True,
        timeout=300,
    )


def append_to_results(section: str) -> None:
    with open(ACT_RESULT_FILE, "a") as fh:
        fh.write(section)


def run_test_case(tc: dict) -> bool:
    name = tc["name"]
    expected = tc["expected"]
    expected_line = f"ASSIGNED_LABELS: {expected}"

    print(f"\n{'='*60}")
    print(f"TEST: {name}")
    print(f"Expected: {expected_line}")
    print(f"{'='*60}")

    with tempfile.TemporaryDirectory() as tmpdir:
        setup_temp_repo(tmpdir, tc["input"])
        result = run_act(tmpdir)

    output = result.stdout + result.stderr

    delimiter = f"\n{'='*60}\n"
    section = (
        f"{delimiter}"
        f"TEST CASE: {name}\n"
        f"EXIT CODE: {result.returncode}\n"
        f"{delimiter}"
        f"{output}"
        f"{delimiter}"
    )
    append_to_results(section)
    print(output[-3000:] if len(output) > 3000 else output)

    passed = True

    if result.returncode != 0:
        print(f"FAIL [{name}]: act exited {result.returncode}")
        passed = False

    if "Job succeeded" not in output:
        print(f"FAIL [{name}]: 'Job succeeded' not found in output")
        passed = False

    if expected_line not in output:
        print(f"FAIL [{name}]: '{expected_line}' not found in output")
        passed = False

    if passed:
        print(f"PASS [{name}]")

    return passed


def main() -> None:
    # Initialise (or clear) act-result.txt
    with open(ACT_RESULT_FILE, "w") as fh:
        fh.write("ACT TEST RESULTS — PR Label Assigner\n")
        fh.write("=" * 60 + "\n")

    results = []
    for tc in TEST_CASES:
        results.append(run_test_case(tc))

    # Summary
    total = len(results)
    passed = sum(results)
    summary = f"\nOVERALL: {passed}/{total} test cases passed\n"
    if passed == total:
        summary += "ALL PASSED\n"
    else:
        failed = [tc["name"] for tc, ok in zip(TEST_CASES, results) if not ok]
        summary += f"FAILED: {', '.join(failed)}\n"

    append_to_results(summary)
    print(summary)
    sys.exit(0 if passed == total else 1)


if __name__ == "__main__":
    main()
